subsys/bluetooth/mesh/adv_multi.c

```diff
@@ -0,0 +1,452 @@
+		return err;
+	}
+
+	err = bt_le_ext_adv_set_data(adv_ins->instance, ad, ad_len, sd, sd_len);
+	if (err) {
+		BT_ERR("Failed setting adv data: %d", err);
+		atomic_clear_bit(adv_ins->flags, ADV_FLAG_ACTIVE);
+		return err;
+	}
+
+	adv_ins->timestamp = k_uptime_get();
+
+	err = bt_le_ext_adv_start(adv_ins->instance, start);
+	if (err) {
+		BT_ERR("Advertising failed: err %d", err);
+		atomic_clear_bit(adv_ins->flags, ADV_FLAG_ACTIVE);
+	}
+
+	return err;
+}
+
+static int buf_send(struct ext_adv *adv_ins, struct net_buf *buf)
+{
+	struct bt_le_ext_adv_start_param start = {
+		.num_events =
+			BT_MESH_TRANSMIT_COUNT(BT_MESH_ADV(buf)->xmit) + 1,
+	};
+	uint16_t duration, adv_int;
+	struct bt_data ad;
+	int err;
+
+	adv_int = MAX(ADV_INT_FAST_MS,
+		      BT_MESH_TRANSMIT_INT(BT_MESH_ADV(buf)->xmit));
+	/* Upper boundary estimate: */
+	duration = start.num_events * (adv_int + 10);
+
+	BT_DBG("type %u len %u: %s", BT_MESH_ADV(buf)->type,
+	       buf->len, bt_hex(buf->data, buf->len));
+	BT_DBG("count %u interval %ums duration %ums",
+	       BT_MESH_TRANSMIT_COUNT(BT_MESH_ADV(buf)->xmit) + 1, adv_int,
+	       duration);
+
+	ad.type = bt_mesh_adv_type[BT_MESH_ADV(buf)->type];
+	ad.data_len = buf->len;
+	ad.data = buf->data;
+
+	/* Only update advertising parameters if they're different */
+	if (adv_param.interval_min != BT_MESH_ADV_SCAN_UNIT(adv_int)) {
+		adv_param.interval_min = BT_MESH_ADV_SCAN_UNIT(adv_int);
+		adv_param.interval_max = adv_param.interval_min;
+	}
+
+	err = adv_start(adv_ins, &adv_param, &start, &ad, 1, NULL, 0);
+	if (!err) {
+		adv_ins->buf = net_buf_ref(buf);
+	}
+
+	bt_mesh_adv_send_start(duration, err, BT_MESH_ADV(buf));
+
+	return err;
+}
+
+static void send_local_adv(struct k_work *work)
+{
+	struct ext_adv *adv_ins = CONTAINER_OF(work, struct ext_adv, work.work);
+	struct net_buf *buf;
+	int err;
+
+	atomic_clear_bit(adv_ins->flags, ADV_FLAG_SCHEDULED);
+
+#if defined(CONFIG_BT_MESH_RELAY)
+	while ((buf = bt_mesh_adv_buf_get(K_NO_WAIT))) {
+#else /* !CONFIG_BT_MESH_RELAY */
+	while ((buf = net_buf_get(&bt_mesh_adv_queue, K_NO_WAIT))) {
+#endif /* CONFIG_BT_MESH_RELAY */
+		/* busy == 0 means this was canceled */
+		if (!BT_MESH_ADV(buf)->busy) {
+			net_buf_unref(buf);
+			continue;
+		}
+
+		BT_MESH_ADV(buf)->busy = 0U;
+		err = buf_send(adv_ins, buf);
+
+		net_buf_unref(buf);
+
+		if (!err) {
+			return; /* Wait for advertising to finish */
+		}
+	}
+}
+
+static void send_relay_adv(struct k_work *work)
+{
+	struct ext_adv *adv_ins = CONTAINER_OF(work, struct ext_adv, work.work);
+	struct net_buf *buf;
+	int err;
+
+	atomic_clear_bit(adv_ins->flags, ADV_FLAG_SCHEDULED);
+
+	while ((buf = net_buf_get(&bt_mesh_relay_queue, K_NO_WAIT))) {
+		/* busy == 0 means this was canceled */
+		if (!BT_MESH_ADV(buf)->busy) {
+			net_buf_unref(buf);
+			continue;
+		}
+
+		BT_MESH_ADV(buf)->busy = 0U;
+		err = buf_send(adv_ins, buf);
+
+		net_buf_unref(buf);
+
+		if (!err) {
+			return; /* Wait for advertising to finish */
+		}
+	}
+}
```

Comment:
It should be possible to get rid of this duplication between `send_local_adv` and `send_relay_adv`. 

 Could for instance change the `tag` to a bitfield enum (that is, ADV_LOCAL is `BIT(0)` and ADV_RELAY is `BIT(1)`), then make the `bt_mesh_adv_buf_get` function take a `tag` parameter, which fetches from one or both of the queues based on the indicated tags.

This would also remove the last remaining reference to `bt_mesh_relay_queue`and `bt_mesh_local_queue` outside the adv.c file, and they would no longer have to be `extern`ed.